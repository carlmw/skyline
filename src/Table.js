import React from 'react';
import PropTypes from 'prop-types';

const renderRow = ({ id, cells }) => (
  <tr key={id}>{cells.map((cell, i) => <td key={i}>{cell}</td>)}</tr>
);

const CitiesTable = ({ loading = false, error = null, data = null }) => {
  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>{error.toString()}</div>;
  }

  if (!data) {
    return <div>No results to display</div>;
  }

  const { columns, rows } = data;

  return (
    <table>
      <thead>
        <tr>
          {columns.map((column, i) => <th key={i}>{column}</th>)}
        </tr>
      </thead>
      <tbody>
        {rows.map(renderRow)}
      </tbody>
    </table>
  );
};
CitiesTable.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.instanceOf(Error),
  data: PropTypes.shape({
    columns: PropTypes.arrayOf(PropTypes.string),
    rows: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        cells: PropTypes.arrayOf(
          PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        ),
      }),
    ),
  }),
};

export default CitiesTable;
