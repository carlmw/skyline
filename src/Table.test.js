import React from 'react';
import { shallow } from 'enzyme';
import Table from './Table';

describe('<Table />', () => {
  const CITY_TABLE_DATA = {
    columns: ['id', 'City', 'Country', 'All Buildings', '100m+'],
    rows: [
      {
        id: 2,
        cells: [2, 'New York City', 'United States', 1390, 803],
      },
    ],
  };

  it('has a loading state', () => {
    const result = shallow(<Table loading/>);

    expect(result).toMatchSnapshot();
  });

  it('has an error state', () => {
    const result = shallow(<Table error={new Error('Failed to load data')} />);

    expect(result).toMatchSnapshot();
  });

  it('displays a no results message when there is no data', () => {
    const result = shallow(<Table />);

    expect(result).toMatchSnapshot();
  });

  it('renders rows when we have data', () => {
    const result = shallow(<Table data={CITY_TABLE_DATA} />);

    expect(result).toMatchSnapshot();
  });
});
