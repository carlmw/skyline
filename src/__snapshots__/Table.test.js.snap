// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> displays a no results message when there is no data 1`] = `
<div>
  No results to display
</div>
`;

exports[`<Table /> has a loading state 1`] = `
<div>
  Loading...
</div>
`;

exports[`<Table /> has an error state 1`] = `
<div>
  Error: Failed to load data
</div>
`;

exports[`<Table /> renders rows when we have data 1`] = `
<table>
  <thead>
    <tr>
      <th
        key="0"
      >
        id
      </th>
      <th
        key="1"
      >
        City
      </th>
      <th
        key="2"
      >
        Country
      </th>
      <th
        key="3"
      >
        All Buildings
      </th>
      <th
        key="4"
      >
        100m+
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="2"
    >
      <td
        key="0"
      >
        2
      </td>
      <td
        key="1"
      >
        New York City
      </td>
      <td
        key="2"
      >
        United States
      </td>
      <td
        key="3"
      >
        1390
      </td>
      <td
        key="4"
      >
        803
      </td>
    </tr>
  </tbody>
</table>
`;
